### PR TITLE
MODTLR-22: Create primary request in borrowing tenant

### PR DESCRIPTION
**Pull request by OleksandrVidinieiev**
## Purpose
Create primary request in borrowing tenant.
Resolves [MODTLR-22](https://folio-org.atlassian.net/browse/MODTLR-22)

## Approach
Extract borrowing tenant ID from access token found in cookies.

#### TODOS and Open Questions
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.

## Learning
_Describe the research stage. Add links to blog posts, patterns, libraries or addons used to solve this problem._

### CODE REVIEW

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 1]
```diff
@@ -84,23 +77,60 @@ public boolean delete(UUID requestId) {
...
+      .requestType(Request.RequestTypeEnum.HOLD)
+      .fulfillmentPreference(Request.FulfillmentPreferenceEnum.HOLD_SHELF);
+  }
+
+  private static void updateEcsTlr(EcsTlr ecsTlr, RequestWrapper primaryRequest,
+    RequestWrapper secondaryRequest) {
 
-    return requestsMapper.mapEntityToDto(ecsTlrRepository.save(
-      requestsMapper.mapDtoToEntity(ecsTlr)));
+    log.info("updateEcsTlr:: updating ECS TLR in memory");
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java:125]:
```suggestion
    log.debug("updateEcsTlr:: updating ECS TLR in memory");
```

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 2]
```diff
@@ -84,23 +77,60 @@ public boolean delete(UUID requestId) {
...
+    ecsTlr.primaryRequestTenantId(primaryRequest.tenantId())
+      .primaryRequestId(primaryRequest.request().getId())
+      .secondaryRequestTenantId(secondaryRequest.tenantId())
+      .secondaryRequestId(secondaryRequest.request().getId())
+      .itemId(secondaryRequest.request().getItemId());
+
+    log.info("updateEcsTlr:: ECS TLR updated in memory");
+    log.debug("updateEcsTlr:: ECS TLR: {}", ecsTlr);
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java:133]:
```suggestion
    log.debug("updateEcsTlr:: ECS TLR: {}", () -> ecsTlr);
```

File: src/main/java/org/folio/service/impl/RequestServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+@Log4j2
+public class RequestServiceImpl implements RequestService {
+  private final TenantScopedExecutionService tenantScopedExecutionService;
+  private final CirculationClient circulationClient;
+  private final UserService userService;
+
+  @Override
+  public RequestWrapper createPrimaryRequest(Request request, String borrowingTenantId) {
+    final String requestId = request.getId();
+    log.info("createPrimaryRequest:: creating primary request {} in borrowing tenant {}",
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/impl/RequestServiceImpl.java:31]:
```suggestion
    log.debug("createPrimaryRequest:: creating primary request {} in borrowing tenant {}",
```

File: src/main/java/org/folio/service/impl/RequestServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    log.debug("createPrimaryRequest:: primary request: {}", () -> primaryRequest);
+
+    return new RequestWrapper(primaryRequest, borrowingTenantId);
+  }
+
+  @Override
+  public RequestWrapper createSecondaryRequest(Request request, String borrowingTenantId,
+    Collection<String> lendingTenantIds) {
+
+    log.info("createSecondaryRequest:: attempting to create secondary request in one of potential " +
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/impl/RequestServiceImpl.java:46]:
```suggestion
    log.debug("createSecondaryRequest:: attempting to create secondary request in one of potential " +
```

File: src/main/java/org/folio/service/impl/RequestServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    String errorMessage = format(
+      "Failed to create secondary request for instance %s in all potential lending tenants: %s",
+      request.getInstanceId(), lendingTenantIds);
+    log.error("createSecondaryRequest:: {}", errorMessage);
+    throw new RequestCreatingException(errorMessage);
+  }
+
+  private RequestWrapper createSecondaryRequest(Request request, String lendingTenantId) {
+    final String requestId = request.getId();
+    log.info("createSecondaryRequest:: creating secondary request {} in lending tenant {}",
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/impl/RequestServiceImpl.java:76]:
```suggestion
    log.debug("createSecondaryRequest:: creating secondary request {} in lending tenant {}",
```

File: src/main/java/org/folio/exception/ResourceNotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package org.folio.exception;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+@Getter
```
[the remark this answers is not in the document]
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/exception/ResourceNotFoundException.java]: I removed this exception, it is not used any more

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java (hunk 1 above)
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java:125]: I'd leave it with info level

File: src/main/java/org/folio/service/impl/RequestServiceImpl.java (hunk 3 above)
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/service/impl/RequestServiceImpl.java:31]: I'd leave it with info level

File: src/main/java/org/folio/service/impl/RequestServiceImpl.java (hunk 4 above)
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/service/impl/RequestServiceImpl.java:46]: I'd leave it with info level

File: src/main/java/org/folio/service/impl/RequestServiceImpl.java (hunk 5 above)
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/service/impl/RequestServiceImpl.java:76]: I'd leave it with info level

File: src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java [hunk 7]
```diff
@@ -42,24 +40,19 @@ public Optional<EcsTlr> get(UUID id) {
...
-      try {
-        return createRequest(ecsTlr, tenantId);
-      } catch (Exception e) {
-        log.error("create:: failed to create a request for tenant {}: {}", tenantId, e.getMessage());
-        log.debug("create:: ", e);
-      }
-    }
-    throw new RequestCreatingException(format(
-      "Failed to create a request for instanceId %s in tenants %s", instanceId, tenantIds));
+    final String requesterId = ecsTlr.getRequesterId();
```
[the remark this answers is not in the document]
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/service/impl/EcsTlrServiceImpl.java]: yes, inlined

**Review by roman-barannyk — APPROVED** [review]

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**